src/wallet/spend.cpp

```diff
@@ -692,7 +708,24 @@ util::Result<SelectionResult> AutomaticCoinSelection(const CWallet& wallet, Coin
         }
 
         // Group outputs and map them by coin eligibility filter
-        FilteredOutputGroups filtered_groups = GroupOutputs(wallet, available_coins, coin_selection_params, ordered_filters);
+        std::vector<OutputGroup> discarded_groups;
+        FilteredOutputGroups filtered_groups = GroupOutputs(wallet, available_coins, coin_selection_params, ordered_filters, discarded_groups);
+
+        // Check if we still have enough balance after applying filters (some coins might be discarded)
+        CAmount total_discarded = 0;
+        CAmount total_unconf_long_chain = 0;
+        for (const auto& group : discarded_groups) {
+            total_discarded += group.GetSelectionAmount();
+            if (group.m_ancestors >= max_ancestors || group.m_descendants >= max_descendants) total_unconf_long_chain += group.GetSelectionAmount();
```

Comment:
1) I think this should respect `fRejectLongChains` flag. If user explicitly asked us to ignore long-chain constraints, then we should respect that.

2) If `fRejectLongChains` flag is set (the default) then the most lenient filter has `max_ancestors/2` and `max_descendants/2`. So this condition could never be hit.

In summary I don't yet understand how one can hit a message described in #23144

Reply:
> 1. I think this should respect `fRejectLongChains` flag. If user explicitly asked us to ignore long-chain constraints, then we should respect that.

It's being respected:
If the user asked to ignore long-chain constraint (`fRejectLongChains=false`) the last filter include the coins, which means that they aren't discarded, so the `discarded_groups` vector is empty, so `total_unconf_long_chain` is zero, so the error is never hit.

> 2. If `fRejectLongChains` flag is set (the default) then the most lenient filter has `max_ancestors/2` and `max_descendants/2`. So this condition could never be hit.

At the contraire, the unconfirmed coins are inside the `available_coins` struct (the user sees them as spendable), and because of the not inclusion of the last filter (due the `fRejectLongChains=true` flag), they are excluded from the created output groups and included inside the discarded groups vector, which is the one used to hit this condition.

> In summary I don't yet understand how one can hit the error described in #23144

Thinks in this way:

If the wallet reject long chains (default), the process discard coins that the user sees as spendable and tries to find a result with the remaining ones. Returning an "Insufficient Funds" if it's not found.
The "discarded groups" vector basically represents that. Coin Selection excluded coins seen as spendable by the user.

Reply:
Yes you're right, I somehow managed to confuse myself about the filters. Thanks for patient explanation.

Still something doesn't sit right. So if a coin is discarded because of `max_ancestors` we give this error message, but if it's discarded by the filter `max_ancestors/2` we just say insufficient funds when the coins are there they just don't meet the filters. There are also other reasons when the coins are discarded, for example unsafe utxos. Even you enable unsafe utxos, we don't allow any chains at all. Why give detailed error message in one case and not the other? Maybe we should just count all discarded coins and generalize the error message a bit?

Reply:
> Still something doesn't sit right. So if a coin is discarded because of max_ancestors we give this error message, but if it's discarded by the filter max_ancestors/2 we just say insufficient funds when the coins are there they just don't meet the filters

no, those coins will be added to the last non-customizable filter which has `max_ancestors-1` and `max_descendants-1`.

> There are also other reasons when the coins are discarded, for example unsafe utxos. Even you enable unsafe utxos, we don't allow any chains at all

The unsafe UTXO filter is just redundant, we are currently filtering unsafe UTXO inside the coins fetching process (`AvailableCoins`).
In other words, when `m_include_unsafe_inputs=false` no unsafe UTXO arrives at Coin Selection.

> Why give detailed error message in one case and not the other? Maybe we should just count all discarded coins and generalize the error message a bit?

The process is already counting all discarded coins and returning early if their exclusion make the available amount go below the target. It's currently returning an "Insufficient Funds", but could say something different too.

I made the too-long-mempool-chain error an special case because this error, alike the untrusted coins that the user can see in a different balance (the unconfirmed/untrusted one), is hidden to the user as it depends on inner mempool policies.
It's not enough for the user to set a flag to spend the unsafe coins, it needs to set another one to craft the tx with an unconfirmed output that surpasses the max descendants policy.

Reply:
I realised my mistake. You can resolve this and ignore my comments. I'll go over the code one more time before I ack